src/framing/Frame.cpp

```diff
@@ -95,78 +95,120 @@ std::ostream& operator<<(std::ostream& os, ErrorCode errorCode) {
   return os << "ErrorCode(" << static_cast<uint32_t>(errorCode) << ")";
 }
 
-std::ostream& operator<<(std::ostream& os, FrameFlags frameFlags) {
-  // TODO Match the Flag names with the AllowedFlags in the Frame declarations
+constexpr const char sFrameFlags_EMPTY[] = "0x00";
+constexpr const char sFrameFlags_METADATA[] = "METADATA";
+constexpr const char sFrameFlags_RESUME_ENABLE[] = "RESUME_ENABLE";
+constexpr const char sFrameFlags_LEASE[] = "LEASE";
+constexpr const char sFrameFlags_KEEPALIVE_RESPOND[] = "KEEPALIVE_RESPOND";
+constexpr const char sFrameFlags_FOLLOWS[] = "FOLLOWS";
+constexpr const char sFrameFlags_COMPLETE[] = "COMPLETE";
+constexpr const char sFrameFlags_NEXT[] = "NEXT";
+
+std::string to_string(
+    FrameFlags frameFlags,
+    const std::vector<std::pair<FrameFlags, std::string>>& allowedFlags) {
+  FrameFlags foundFlags = FrameFlags::EMPTY;
 
   std::stringstream ss;
   std::string delimeter = "";
-  if (!!(frameFlags & FrameFlags::NEXT)) {
-    ss << "NEXT";
-    delimeter = "|";
-  }
-  if (!!(frameFlags & FrameFlags::COMPLETE)) {
-    ss << delimeter << "COMPLETE";
-    delimeter = "|";
-  }
-  if (!!(frameFlags & FrameFlags::FOLLOWS)) {
-    ss << delimeter << "FOLLOWS";
-    delimeter = "|";
+  for (auto& pair : allowedFlags) {
+    if (!!(frameFlags & pair.first)) {
+      ss << delimeter << pair.second;
+      delimeter = "|";
+      foundFlags |= pair.first;
+    }
   }
-  if (!!(frameFlags & FrameFlags::METADATA)) {
-    ss << delimeter << "METADATA";
-    delimeter = "|";
-  }
-  if (!!(frameFlags & FrameFlags::IGNORE)) {
-    ss << delimeter << "IGNORE";
-    delimeter = "|";
-  }
-  if (!delimeter.empty()) {
-    return os << ss.str();
+
+  if (foundFlags != frameFlags) {
+    ss << frameFlags;
+  } else if (delimeter.empty()) {
+    ss << sFrameFlags_EMPTY;
   }
-  return os << "EMPTY";
+  return ss.str();
+}
+
+std::ostream& operator<<(std::ostream& os, FrameFlags frameFlags) {
+  std::bitset<16> flags(static_cast<uint16_t>(frameFlags));
+  return os << flags;
 }
 
 std::ostream& operator<<(std::ostream& os, const FrameHeader& header) {
   return os << header.type_ << "[" << header.flags_ << ", " << header.streamId_
             << "]";
 }
+
+std::ostream& operator<<(
+    std::ostream& os,
+    const std::pair<
+        const FrameHeader&,
+        const std::vector<std::pair<FrameFlags, std::string>>&>&
```

Comment:
This is really rough to understand...  The second element in this pair is a `vector<pair<Flags, std::string>>&` which is also inside a wrapping pair<...,...>&.  All for the purpose of fitting everything into a single argument for `operator<<`.

There's no need to use vectors to track which flags are allowed for which frame header.  You can write a `FrameFlags allowedFlags(FrameType);` function and that can use the type in the header.  Then have a `const char* toString(FrameFlags, FrameType);` that'll decide which string to return (need the type to differentiate between flags like `FOLLOWS` and `KEEPALIVE_RESPOND`).

I think instead of writing
`std::ostream& operator<<(std::ostream&, const std::vector<std::pair<FrameFlags, std::string>>&>&);`
I think we can do everything we want with
`std::ostream& operator<<(std::ostream&, const FrameHeader&);`